modules/explorer-core/src/main/scala/org/ergoplatform/explorer/protocol/dex.scala

```diff
@@ -2,43 +2,137 @@ package org.ergoplatform.explorer.protocol
+  private val BuyContractTokenAmountIndexInConstants = 8
 
-  /** template for the buyer contract as of http://github.com/ScorexFoundation/sigmastate-interpreter/blob/42e55cbfd093252b8005e4607970764dd6610cbe/contract-verification/src/main/scala/sigmastate/verification/contract/AssetsAtomicExchange.scala#L33-L45
-    */
-  val sellContractTemplate: HexString = HexString(
-    refineMV[HexStringSpec](
-      "eb027300d1eded91b1a57301e6c6b2a5730200040ed801d60193e4c6b2a5730300040ec5a7eded92c1b2a57304007305720193c2b2a5730600d07307"
-    )
-  )
+  def sellContractInstance[F[_]: ContravariantRaise[
+    *[_],
+    DexContractInstantiationFailed
+  ]: Applicative](
+    tokensPrice: Long
+  ): F[ErgoTree] = {
+    import org.ergoplatform.sigma.verified.VerifiedTypeConverters._
+    // since we're only using compiled contracts for constant(parameters) extraction PK value does not matter
+    Try {
+      val anyPk = sigmastate.eval.SigmaDsl.SigmaProp(ProveDlog(constants.group.generator))
+      AssetsAtomicExchangeCompilation.sellerContractInstance(tokensPrice, anyPk).ergoTree
+    }.toEither
+      .leftMap(e => DexContractInstantiationFailed(e.getMessage))
+      .toRaise
+  }
 
-  // assuming seller contract from http://github.com/ScorexFoundation/sigmastate-interpreter/blob/42e55cbfd093252b8005e4607970764dd6610cbe/contract-verification/src/main/scala/sigmastate/verification/contract/AssetsAtomicExchange.scala#L33-L45
-  val BuyContractTokenIdIndexInConstants     = 6
-  val BuyContractTokenAmountIndexInConstants = 8
+  def sellContractTemplate[F[_]: ContravariantRaise[*[_], DexContractInstantiationFailed]: ContravariantRaise[
+    *[_],
+    ErgoTreeDeserializationFailed
+  ]: ContravariantRaise[
+    *[_],
+    ErgoTreeSerializationFailed
+  ]: ContravariantRaise[
+    *[_],
+    RefinementFailed
+  ]: Monad]: F[HexString] =
+    // parameter values does not matter, we're extracting ErgoTree template (with placeholders in places of values)
+    sellContractInstance[F](0L)
+      .flatMap(ergoTreeTemplateBytes[F])
+      .flatMap(bytes => HexString.fromString(Base16.encode(bytes)))
 
-  /** template for the buyer contract as of http://github.com/ScorexFoundation/sigmastate-interpreter/blob/42e55cbfd093252b8005e4607970764dd6610cbe/contract-verification/src/main/scala/sigmastate/verification/contract/AssetsAtomicExchange.scala#L12-L32
-    */
-  val buyContractTemplate: HexString = HexString(
-    refineMV[HexStringSpec](
-      "eb027300d1eded91b1a57301e6c6b2a5730200040ed803d601e4c6b2a5730300020c4d0ed602eded91b172017304938cb27201730500017306928cb27201730700027308d60393e4c6b2a5730900040ec5a7eded720293c2b2a5730a00d0730b7203"
+  def buyContractInstance[F[_]: ContravariantRaise[
+    *[_],
+    DexContractInstantiationFailed
+  ]: ContravariantRaise[
+    *[_],
+    Base16DecodingFailed
+  ]: Monad](
+    tokenId: TokenId,
+    tokenAmount: Long
+  ): F[ErgoTree] =
+    stringBase16ToBytes(tokenId.value).flatMap { tokenIdBytes =>
```

Comment:
I think `tokenId.value` should't be just string (initially I thought it would never be decoded). Let me wrap it into `HexString`.

Reply:
Makes sense. Then `Base16DecodingFailed` can be reverted to `HexString`(`String` was needed only for decoding token id).

Reply:
Done. Please update your branch accordingly)

Reply:
@oskin1 Thanks! Updated.